keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,983 @@
+stepping through it.
+
+He therefore wants to enable logging so that `go test` shows detailed log
+output for a failed test case, and only for that test case. He wants to run it
+like that by default in the CI so that when the problem occurs, all information
+is immediately available. This is important because the error might not show up
+when trying the same test invocation locally.
+
+When everything works, he wants `go test` to hide the log output to avoid
+blowing up the size of the CI log files.
+
+For each inner test case he adds a `NewTestContext(t)` invocation and uses the
+returned context and logger for that test case.
+
+#### Story 4
+
+Client developer Joan wants to use client-go in his application, but is [less
+interested in log messages from
+it](https://kubernetes.slack.com/archives/CG3518SFJ/p1634217685020400). Joan
+makes the log messages from client-go less verbose by creating a logger with
+`logger.V(1)` and passing that to the client-go code. Now a `logger.V(3).Info`
+call in client-go is the same as a `logger.V(4).Info` in the application and
+will not be shown for `-v=3`.
+
+
+### Risks and Mitigations
+
+#### Design
+
+##### Uninitialized logger
+
+One risk is that code uses an uninitialized `logr.Logger`, which would lead to
+nil pointer panics. This gets mitigated with the logr convention that a
+`logr.Logger` passed by value always must be usable. When the logger is optional,
+this needs to be indicated by passing a pointer.
+
+##### Logging during initialization
+
+Another problem is that log messages during program startup need to be handled
+somehow. We then can either:
+- discard the message, which would break the current `klog.Error(...);
+  os.Exit(1)` code pattern when an error occurs during initialization log
+- log with some default configuration (for example, as text)
+
+In both cases we have the problem that a logger retrieved by some init code
+before logging initialization will continue to use that behavior (discard or
+log with default configuration) even after logging gets initialized.
+
+Ideally, no log messages should be emitted before the program is done with
+setting up logging as intended. This is already [problematic
+now](https://github.com/kubernetes/kubernetes/issues/100152) because output may
+change from text (the current klog default) to JSON (once initialized). There
+will be no automatic mitigation for this. Such log calls will have to be found
+and fixed manually, for example by [passing the error back to `main`
+instead](https://github.com/kubernetes/kubernetes/pull/104774), which is part
+of an effort to [remove the dependency on logging before an unexpected program
+exit](https://github.com/kubernetes/kubernetes/issues/102231).
+
+To support identifying log messages that need to be fixed, the default logger
+during program startup will be using the klog text output instead of discarding
+the output and a `INIT-LOG-FIXME` prefix will be inserted into the log output.
```

Comment:
Perhaps they should freak out and file an issue against the component which emits such a log message?

But I'm fine with toning it down to just "INIT-LOG". Let's hear what @serathius thinks.

Reply:
Both `INIT-LOG` and `INIT-LOG-FIXME` doesn't seem actionable for neither user or contributor. If possible I would prefer more explicit message lsimilar to what glog does. https://github.com/golang/glog/blob/master/glog.go#L680

If glog is unititalized it will write prefix `ERROR: logging before flag.Parse: ` to each line. Could we do something like `logging before logger init`, still this is more cosmetic change. 
Overall approach makes sense

Reply:
Both `INIT-LOG` and `INIT-LOG-FIXME` doesn't seem actionable for neither user or contributor. If possible I would prefer more explicit message lsimilar to what glog does. https://github.com/golang/glog/blob/master/glog.go#L680

If glog is unititalized it will write prefix `ERROR: logging before flag.Parse: ` to each line. Could we do something like `logging before logger init`, still this is more cosmetic change. 

Overall approach makes sense

Reply:
So now we don't have a decision between `INIT-LOG` or `INIT-LOG-FIXME` but rather a third option on the table :sweat_smile: 

I think the "actionable" part is covered by the error message that will get emitted once when the final logger is set and some messages were printed earlier. That message is documented further down in the `SetFallbackLogger` proposal. That is very verbose and provides instructions. Therefore we don't need a long text in front of each init log entry, just a unique marker that people can grep for.

Let's go with `INIT-LOG` for the prefix. I've also updated this section here with more information about that error message, that was missing.